### ISSUE TITLE
Difficult to override the `Cabal` version used for setup.hs

### ISSUE THREAD

**Issue by shlevy** [NixOS/nixpkgs#46166]
Unlike other dependencies of setup, `Cabal` is not passed as an argument to packages that need it (i.e., all of them). As such, it becomes difficult to override the version Cabal to use. You can override `setupHaskellDepends` for a specific library, but if you want to apply it to all haskell packages you need to do some gymnastics to avoid infinite recursion.

**Comment by shlevy** [comment]
@Ericson2314 @peti @basvandijk @ryantm

**Comment by matthewbauer** [comment]
Would https://github.com/NixOS/nixpkgs/pull/43300 fix this?

**Comment by shlevy** [comment]
I'm not sure... But I don't think so. In that world, compiling setup.hs would still use the bundled Cabal with ghc by default.

**Comment by matthewbauer** [comment]
Well this would make overlays work correctly in this case. Without a global package db, you should be able to override Cabal directly like:

```
self: super: {
  Cabal = super.Cabal_2_4_0_0;
}
```

Unfortunately, the flag to do this, ```-no-global-package-db``` is broken in Cabal apparently (https://github.com/haskell/cabal/issues/5445).

**Comment by shlevy** [comment]
How would that work? Packages currently don't depend on `Cabal` as an argument.

**Comment by matthewbauer** [comment]
I think you're right that this won't fix it 100% but I think it would make it possible to pretty easily override the Cabal version used in the generic builder. ```-no-global-package-db``` should make it possible to just add your custom mkDerivation that always add this:

```
self: super: {
  mkDerivation = args: super.mkDerivation (args // {
    setupHaskellDepends = (args.setupHaskellDepends or []) ++ Cabal_2_4_0_0;
  });
}
```

You just have to ensure that the right Cabal is pulled in (something that I think would have to be addressed in that PR I mentioned).

**Comment by shlevy** [comment]
Right, but then you run into infinite recursion issues where you have to explicitly filter out your custom Cabal and its dependencies

**Comment by shlevy** [comment]
It's not clear why the other PR is relevant. We could do that today, just defaulting to null in absence of a specific override

**Comment by Ericson2314** [comment]
That's true. There's just the slight sanity check of not getting both the built-in and custom Cabal with that approved. I vaguely recall having some issues with that in the past.
